### PR TITLE
update Readme to document Future, async and await

### PR DESCRIPTION
**Pull request by SimonLab**
ref: #9

Add an `asynchronous` section, to explain how to use the `Future` class and the `async/await` and `then` keywords

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -194,6 +194,83 @@ dynamic int a; // https://repl.it/repls/GreenDeadMatch
 a = 42;
 ```
 
+## Asynchronous events
+
+Dart provides the `Future` class to represent asynchronous events.
+For example the following `hello` function will return a `String`  in a near futur:
```
Comment (nelsonic) [inline README.md]:
```suggestion
For example the following `hello` function will return a `String`  in a near future: 
```

File: README.md [hunk 2]
```diff
@@ -194,6 +194,83 @@ dynamic int a; // https://repl.it/repls/GreenDeadMatch
...
+Dart provides the `Future` class to represent asynchronous events.
+For example the following `hello` function will return a `String`  in a near futur: 
+
+```dart
+Future<String> hello() { // The type parameter of Future is a String, represented by <String>
+    return Future.delayed(Duration(seconds: 2), () => "hello"); // Using the Future.delayed function to create a time gap of 2 seconds
+}
+```
+
+A function with a retun type of `Future<T>` (`T` here represent any types) can have two state returns `Uncomplete` and `Complete`.
```
Comment (nelsonic) [inline README.md]:
```suggestion
A function with a return type of `Future<T>` (where `T` represents any type) 
can have _three_ possible state returns: 
`Uncompleted`, `Completed ` (_with_) ***success*** and `Completed` (_with_) ***error***
```
See: https://medium.com/@saad.eloulladi/flutter-dart-what-are-futures-2f9eea158d70

File: README.md [hunk 3]
```diff
@@ -194,6 +194,83 @@ dynamic int a; // https://repl.it/repls/GreenDeadMatch
...
+When completed the function will return the type `T`, in our case the `hello` function returns a `String`.
+If an error occurs then the function will return an `Error`.
+
+To be able to use the returned value of an asynchronus function we can use the `async` and `await` keywords.
+We need first to describe the function using an asynchronous function by adding the `async` keyword before the body of the function.
+Then we using the asynchronous function we prefix the call to the function with `await`. This will stop the process of the function
+and wait for the futur result. For example we can create a `main` function which will use our `hello` function:
```
Comment (nelsonic) [inline README.md]:
```suggestion
and wait for the future result. For example we can create a `main` function which will use our `hello` function:
```

File: README.md [hunk 4]
```diff
@@ -194,6 +194,83 @@ dynamic int a; // https://repl.it/repls/GreenDeadMatch
...
+
+```dart
+void main() async {
+    String hi = await hello();
+    print(hi); // print "hello" after 2 seconds
+}
+
+Future<String> hello() { // The type parameter of Future is a String, represented by <String>
+    return Future.delayed(Duration(seconds: 2), () => "hello"); // Using the Future.delayed function to create a time gap of 2 seconds
```
Comment (nelsonic) [inline README.md]:
```suggestion
    return Future.delayed(Duration(seconds: 2), () => "hello"); // Use Future.delayed to delay execution by 2 sec.
```

File: README.md [hunk 5]
```diff
@@ -194,6 +194,83 @@ dynamic int a; // https://repl.it/repls/GreenDeadMatch
...
+}
+```
+
+To test this code you can copy/paste it and run it on dartpad: https://dartpad.dev/
```
Comment (nelsonic) [inline README.md]: 👍

File: README.md [hunk 6]
```diff
@@ -194,6 +194,83 @@ dynamic int a; // https://repl.it/repls/GreenDeadMatch
...
+    dynamic hi = hello();
+    print(hi); // print "Instance of _Future<String>"
+}
+
+Future<String> hello() { // The type parameter of Future is a String, represented by <String>
+    return Future.delayed(Duration(seconds: 2), () => "hello"); // Using the Future.delayed function to create a time gap of 2 seconds
+}
+```
+
+We can also use the `then` function which take a callback function to get the futur value:
```
Comment (nelsonic) [inline README.md]:
```suggestion
We can also use the `then` function which takes a callback function to get the future value:
```

File: README.md [hunk 7]
```diff
@@ -194,6 +194,83 @@ dynamic int a; // https://repl.it/repls/GreenDeadMatch
...
+        print(futureValue); // print after 2 seconds "hello"
+    });
+}
+
+Future<String> hello() { // The type parameter of Future is a String, represented by <String>
+    return Future.delayed(Duration(seconds: 2), () => "hello"); // Using the Future.delayed function to create a time gap of 2 seconds
+}
+```
+
+Compare to `await`, `then` will not stop the process of the function and continue the execution:
```
Comment (nelsonic) [inline README.md]:
```suggestion
Compared to `await`, `then` will not stop the process of the function and continue the execution:
```

**Comment by SimonLab** [comment]
Thanks for the reviews! @nelsonic I've merged your suggestions and it should be ready to merge now :+1:

**Comment by nelsonic** [comment]
@SimonLab Thanks for accepting the suggestions.
Merging. 👍

**Pull request merged**